Userland/Libraries/LibVirtGPU/CommandBufferBuilder.cpp

```diff
@@ -0,0 +1,310 @@
+        // Pad end with null bytes
+        memset(&dest[string.length()], 0, 4 * num_required_words - string.length());
+    }
+    void finalize()
+    {
+        if (!m_finalized) {
+            m_finalized = true;
+            size_t num_elems = m_buffer.size() - m_start_offset - 1;
+            m_buffer[m_start_offset] = encode_command(num_elems, m_command_mid, m_command);
+        }
+    }
+    ~CommandBuilder()
+    {
+        if (!m_finalized)
+            finalize();
+    }
+
+private:
+    Vector<u32>& m_buffer;
+    size_t m_start_offset;
+    Protocol::VirGLCommand m_command;
+    u32 m_command_mid;
+    bool m_finalized { false };
+};
+
+void CommandBufferBuilder::append_set_tweaks(u32 id, u32 value)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::SET_TWEAKS, 0);
+    builder.appendu32(id);
+    builder.appendu32(value);
+}
+
+void CommandBufferBuilder::append_transfer3d(Protocol::ResourceID resource, size_t width, size_t height, size_t depth, size_t direction)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::TRANSFER3D, 0);
+    builder.appendu32(resource.value()); // res_handle
+    builder.appendu32(0);                // level
+    builder.appendu32(242);              // usage
+    builder.appendu32(0);                // stride
+    builder.appendu32(0);                // layer_stride
+    builder.appendu32(0);                // x
+    builder.appendu32(0);                // y
+    builder.appendu32(0);                // z
+    builder.appendu32(width);            // width
+    builder.appendu32(height);           // height
+    builder.appendu32(depth);            // depth
+    builder.appendu32(0);                // data_offset
+    builder.appendu32(direction);        // direction
+}
+
+void CommandBufferBuilder::append_end_transfers_3d()
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::END_TRANSFERS, 0);
+}
+
+void CommandBufferBuilder::append_draw_vbo(Protocol::PipePrimitiveTypes primitive_type, u32 count)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::DRAW_VBO, 0);
+    builder.appendu32(0);                             // start
+    builder.appendu32(count);                         // count
+    builder.appendu32(to_underlying(primitive_type)); // mode
+    builder.appendu32(0);                             // indexed
+    builder.appendu32(1);                             // instance_count
+    builder.appendu32(0);                             // index_bias
+    builder.appendu32(0);                             // start_instance
+    builder.appendu32(0);                             // primitive_restart
+    builder.appendu32(0);                             // restart_index
+    builder.appendu32(0);                             // min_index
+    builder.appendu32(0xffffffff);                    // max_index
+    builder.appendu32(0);                             // cso
+}
+
+void CommandBufferBuilder::append_gl_clear(float r, float g, float b, float a)
```

Comment:
It seems weird to leak `GLContext` terminology (`gl_clear`) into `LibVirtGPU`. Maybe just `append_clear` for this method and the one below?